python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,480 @@
+                <float*> row_ptr,
+                <int*> smp_ptr,
+                <int> self.nsamples_random,
+                <int> maxsample,
+                <uint64_t> random_state)
+
+        # # evaluate model on combinations
+
+        if self.model_gpu_based:
+            self.y = self.model(self.synth_data)
+        else:
+            try:
+                self.y = cp.array(self.model(
+                    self.synth_data.to_output('numpy'))
+                )
+            except TypeError:
+                raise TypeError('Explainer can only explain models that can '
+                                'take GPU data or NumPy arrays as input.')
+
+        y_hat = cp.mean(
+            cp.array(self.y).reshape((self.nsamples,
+                                      self.background.shape[0])),
+            axis=1
+        )
+
+        # todo: minor optimization can be done by avoiding this array
+        # if l1 reg is not needed
+        nonzero_inds = cp.arange(self.M)
+
+        # call lasso/lars if needed
+        if l1_reg == 'auto':
+            if self.nsamples / self.max_samples < 0.2:
+                nonzero_inds = cp.nonzero(
+                    Lasso(alpha=l1_reg).fit(self.mask, y_hat).coef_
+                )[0]
+                if len(nonzero_inds) == 0:
+                    return cp.zeros(self.M), np.ones(self.M)
+
+        else:
+            if not has_sklearn():
+                raise ImportError("Scikit-learn needed for lars l1 "
+                                  "regularization currently.")
+            else:
+                warn("LARS is not currently GPU accelerated, using "
+                     "Scikit-learn.")
+
+                from sklearn.linear_model import LassoLarsIC, lars_path
+                if (isinstance(l1_reg, str)
+                        and l1_reg.startswith("num_features(")):
+                    r = int(l1_reg[len("num_features("):-1])
+                    nonzero_inds = lars_path(
+                        self.mask, y_hat, max_iter=r)[1]
+                elif (isinstance(l1_reg, str) and l1_reg == "bic" or
+                        l1_reg == "aic"):
+                    nonzero_inds = np.nonzero(
+                        LassoLarsIC(criterion=l1_reg).fit(self.mask,
+                                                          y_hat).coef_)[0]
+
+        # weighted linear regression
+        # todo: see wheter change to use cuML linear regression with weights
+        # todo: small optimizations
+        y_hat = y_hat - self.expected_value
+
+        y_hat = y_hat - self.mask[:, nonzero_inds[-1]] * (
+            self.link_fn(self.fx) - self.link_fn(self.expected_value)
+        )
+
+        etmp = cp.transpose(
+            cp.transpose(
+                self.mask[:,
+                          nonzero_inds[:-1]]) - self.mask[:, nonzero_inds[-1]])
+
+        tmp = cp.dot(cp.dot(etmp.T, cp.diag(self.weights)), etmp)
+
+        # cupy linalg solve requires tmp to be square and full rank,
```

Comment:
not giving up on this one ;)
the same conditions that make cp.solve(tmp, X'y) fail will also make cp.linalg.inv fail...

also, I'd recommend breaking the whole weighted cupy ols out to a separate function maybe with its own separate tests.

Reply:
we should probably add a nice lightweight fully tested, weighted, on-gpu ols anyways in a future rev